### PR TITLE
Add zip methods to appium support

### PR DESCRIPTION
**Pull request by dpgraham**
* Uses `extract-zip` and `yauzl` (yauzl is dependency of extract-zip) to extract zip files
* Uses `archiver` to create zip files
* Added methods `extractAllTo`, `readEntries` and `toInMemoryZip` to the API
* Next step is to refactor out other repos that use `adm-zip` and use these methods instead

(https://github.com/appium/appium/issues/7409)

### CODE REVIEW

File: lib/zip.js [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+ * @param {str} destDir 
+ */
+async function _extractEntryTo (zipfile, entry, destDir) {
+  // Create directory if doesn't exist already
+  const relDir = path.dirname(entry.fileName);
+  await mkdirp(path.resolve(destDir, relDir));
+  
+  // Stream the contents of the entry to a file
+  return new B((resolve) => {
+    zipfile.openReadStream(entry, (err, readStream) => {
```
Comment (imurchie) [inline lib/zip.js]: Do we not want to check the error here, and `reject` the promise if necessary?

File: lib/zip.js [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+      const writeStream = createWriteStream(path.resolve(destDir, entry.fileName), {flags: 'w'});  
+      readStream.pipe(writeStream);
+      readStream.on('end', resolve);
```
Comment (imurchie) [inline lib/zip.js]: We might also want error handling on the stream.

File: lib/zip.js [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+async function toInMemoryZip (srcDir) {
+  const tempOutputDir = await tempDir.openDir();
+  return new B((resolve, reject) => {
+    // Create a writable stream that zip is streamed to 
+    const output = createWriteStream(path.resolve(tempOutputDir, 'temp.zip'));
+    output.on('finish', async () => {
+      try {
+        const zipContents = await fs.readFile(path.resolve(tempOutputDir, 'temp.zip'));
+        resolve(zipContents);
+      } catch (err) { reject(err); }
```
Comment (imurchie) [inline lib/zip.js]: Put the `reject` on a new line.

File: lib/zip.js [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+    });
+
+    // Start streaming zip data to the above writable stream
+    const archive = archiver('zip', {
+      zlib: { level: 9 }
```
Comment (imurchie) [inline lib/zip.js]: No spaces around object literal.

File: lib/zip.js [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+ * @param {Object} entry 
+ * @param {str} destDir 
+ */
+async function _extractEntryTo (zipfile, entry, destDir) {
+  // Create directory if doesn't exist already
+  const relDir = path.dirname(entry.fileName);
+  await mkdirp(path.resolve(destDir, relDir));
+  
+  // Stream the contents of the entry to a file
+  return new B((resolve) => {
```
Comment (imurchie) [inline lib/zip.js]: `return await new B((resolve) => {`

File: lib/zip.js [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+  });
+}
+
+/**
+ * Converts contents of local directory to an in-memory .zip buffer
+ * @param {*} srcDir 
+ */
+async function toInMemoryZip (srcDir) {
+  const tempOutputDir = await tempDir.openDir();
+  return new B((resolve, reject) => {
```
Comment (imurchie) [inline lib/zip.js]: `return await new B((resolve, reject) => {`

**Comment by dpgraham** [comment]
@imurchie I agree with the confusion of mixing promises and async. I overhauled it to use mostly async and only minimally use promises in cases where we need to bind to EventEmitter events.

e.g.:
```javascript
let promise = new B((resolve, reject) => {
  eventEmitter.on('finished', resolve);
  eventEmitter.on('error', reject);
});
```

I also updated the tests like you asked to force more errors. Good call on that, I spotted some issues with the code when I did that.

**Comment by imurchie** [comment]
Testing FTW!

File: lib/zip.js [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+  const writeStream = createWriteStream(path.resolve(destDir, entry.fileName), {flags: 'w'});  
+  const writeStreamPromise = new B((resolve, reject) => {
+    writeStream.once('finish', resolve);
+    writeStream.once('error', reject);
+  });
+
+  // Create zipReadStream and pipe data to the write stream
+  // (for some odd reason B.promisify doesn't work on zipfile.openReadStream, it causes an error 'closed')
+  const zipReadStream = await new B((resolve, reject) => {
+    zipfile.openReadStream(entry, (err, readStream) => err ? reject(err) : resolve(readStream));
```
Comment (imurchie) [inline lib/zip.js]: Slick!

File: lib/zip.js [hunk 8]
```diff
@@ -0,0 +1,116 @@
...
+  const zipWriteStreamPromise = new B((resolve, reject) => {
+    zipWriteStream.once('finish', resolve);
+    zipWriteStream.once('error', reject);
+  });
+
+  // Zip 'srcDir' and stream it to the above writable stream
+  const archive = archiver('zip', {
+    zlib: {level: 9}    
+  });
+  archive.once('error', (err) => { throw new Error(err); });
```
Comment (imurchie) [inline lib/zip.js]: `err` ought to already be an `Error` object, so re-wrapping it is not necessary.

File: test/zip-specs.js [hunk 9]
```diff
@@ -0,0 +1,115 @@
+import chai from 'chai';
+import chaiAsPromised from 'chai-as-promised';
+import path from 'path';
+import { zip } from '..';
+import { tempDir, fs } from '../index';
```
Comment (imurchie) [inline test/zip-specs.js]: These two are the same import, no? `import { zip, tempDir, fs } from '..';` would suffice.

**Comment by imurchie** [comment]
Published in `appium-support@2.8.0`.

**Pull request merged**